### PR TITLE
feat(eslint-plugin): [no-unused-expressions] extend for optional chaining

### LINKED ISSUE #1138: [no-unused-expressions] False positive with optional chaining
**Repro**

```JSON
{
  "rules": {
    "no-unused-expressions": ["error", { "allowShortCircuit": true, "allowTernary": true, "allowTaggedTemplates": true }],
  }
}
```

```TS
class Test {
  test() {
  }
}
declare const t: Test | undefined
t?.test()
```

**Expected Result**

**Actual Result**
```
D:\test\test.ts
  6:1  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.5.0` |
| `@typescript-eslint/parser`        | `2.5.0` |
| `TypeScript`                       | `3.7.0-dev.20191021` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `v10.15.3` |
| `npm`                              | `6.10.1` |

### PR DESCRIPTION
**Pull request by Nizarius**
Fixes #1138

To note: this PR depends on #1169, so it's better to watch after first one

### CODE REVIEW

**Comment by Nizarius** [comment]
@bradzacher yeah, great idea! I will do it.

**Comment by Nizarius** [comment]
@bradzacher done

File: packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+a() && function namedFunctionInExpressionContext () {f();}
+        `,
+      errors: error([
+        {
+          line: 2,
+          column: 1,
+        },
+      ]),
+    },
+  ],
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts]: You'll need to add some invalid test cases for optional chaining as well.


In particular you need to test cases like:
```ts
a?.b
(a?.b).c

a?.['b']
(a?.['b']).c

a?.b()?.c
(a?.b()).c
```

Have a look at some of the examples in the parser fixtures to get a sense for all of the cases that exist in optional chaining:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-call-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-call.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-element-access-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-element-access.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain.src.ts

Comment (Nizarius) [reply packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts]: Done

**Comment by Nizarius** [comment]
@bradzacher added docs and new tests. Could you explain to me how I should update documentation parsing to avoid errors with 'Incorrect line number for'?

Also, should I add this rule to recommended ones to prevent errors with base configuration?

**Comment by bradzacher** [comment]
> Could you explain to me how I should update documentation parsing to avoid 'Incorrect line number for' errors?

The rule needs to be listed in the plugin readme. The table should be sorted alphabetically.

> Also, should I add this rule to recommended ones to prevent errors with base configuration?

Changes to the recommended config are considered breaking changes, so we cannot do that for now.
Feel free to raise a separate tracking issue so we can keep track of doing this when we start on a 3.0.0 release.

File: packages/eslint-plugin/src/configs/all.json [hunk 2]
```diff
@@ -50,6 +50,7 @@
     "@typescript-eslint/no-unnecessary-qualifier": "error",
     "@typescript-eslint/no-unnecessary-type-arguments": "error",
     "@typescript-eslint/no-unnecessary-type-assertion": "error",
+    "@typescript-eslint/no-unused-expressions": "error",
```
Comment (ulrichb) [inline packages/eslint-plugin/src/configs/all.json]: Can we turn off ESLint's `no-unused-expressions`?

Comment (bradzacher) [reply packages/eslint-plugin/src/configs/all.json]: ecb3f4ec9b9f8b56896b0ad985547de0e6608381

Comment (ulrichb) [reply packages/eslint-plugin/src/configs/all.json]: @bradzacher Cool. Thx!!

**Comment by glen-84** [comment]
@JamesHenry This isn't in the release notes?

**Comment by bradzacher** [comment]
@glen-84 it was released earlier - in 2.7
https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.7.0

**Comment by glen-84** [comment]
🤦‍♂ I must have missed that.

**Pull request merged**